main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala

```diff
@@ -303,14 +323,19 @@ object ParsedAst {
       */
     case class False(sp1: SourcePosition, sp2: SourcePosition) extends ParsedAst.Literal
 
+    sealed trait CharCode {
+      val sp1: SourcePosition
+      val sp2: SourcePosition
```

Comment:
Kind of. The reason is for the truncated code case.

```scala
        case (u @ ParsedAst.CharCode.Escape(sp1, "u", _)) :: rest =>
          val code = rest.takeWhile(_.isInstanceOf[ParsedAst.CharCode.Literal])
          val sp2 = code.lastOption.getOrElse(u).sp2
          WeederError.TruncatedUnicodeEscape(mkSL(sp1, sp2)).toFailure
```

`code`'s type is `List[ParsedAst.CharCode]` because of the signature of `takeWhile`, even though we know that it's really `List[ParsedAst.CharCode.Literal]`
So when we get `sp2` here, we're calling `CharCode.sp2` and not `CharCode.Literal.sp2`.

I tried creating a `collectWhile` function, but that ended up being overengineered and messy.
The other option is just to cast, but I feel sad when I do that.

Reply:
OK, thats legitimate